### ISSUE TITLE
SOAP-ERROR: Parsing WSDL:  Premature end of data in tag definitions

### ISSUE THREAD

**Issue by kev22257** [opendns/autotask-php#163]
Hi,

We recently upgraded our application to use the new version of autotask-php to support the 1.6 API.  Everything had been working well for about a month now, but suddenly we are getting this error trying to look at the auto-discover WSDL.

> PHP Fatal error:  Uncaught SoapFault exception: [WSDL] SOAP-ERROR: Parsing WSDL: Couldn't load from 'https://webservices.autotask.net/atservices/1.6/atws.wsdl' : Premature end of data in tag definitions line 2

I don't think this is specifically related to the autotask-php library, but I hoped the people that might look at the Issues here would have a suggestion.  I spent some time researching the issue online and mostly what I found was a bug in PHP 5.3 related to Content-Length.  We are on PHP 5.4, but I went down the rabbit hole anyways and validated that Autotask seems to be setting that header correctly both when compressing and not compressing the response.  I have tried setting various options on SoapClient, but nothing seems to affect the behavior.

It is failing on our first call to ATWS\Client.  Originally I only had trace enabled, but added other options one by one to see if it affected the behavior: it did not.

Any suggestions?

```
function get_at_client() {

    // source the Autotask library
    require_once(__DIR__ . "/autotask/src/autoload.php");

    // Autotask API username and password
    $username = 'xxxxx';
    $password = 'xxxxx';
    $integration_code = "xxxxx";

    // Autotask login WSDL
    $auth_wsdl = 'https://webservices.autotask.net/atservices/1.6/atws.wsdl';
    $opts = array('trace'        => 1,
                  'compression'  => SOAP_COMPRESSION_ACCEPT | SOAP_COMPRESSION_GZIP,
                  'soap_version' => SOAP_1_2,
                  'cache_wsdl'   => WSDL_CACHE_NONE,
                  'features'     => SOAP_SINGLE_ELEMENT_ARRAYS);
    $client = new ATWS\Client($auth_wsdl, $opts); // Fails here!!!
    $zoneInfo = $client->getZoneInfo($username);

    // Autotask Client and ticket field info for pick lists
    $auth_opts = array(
        'login' => $username,
        'password' => $password,
        'trace' => 1
    );
    $wsdl = str_replace('.asmx', '.wsdl', $zoneInfo->getZoneInfoResult->URL);
    $client = new ATWS\Client($wsdl, $auth_opts, $integration_code);

    return $client;
}

```

**Comment by SirPoot-** [comment]
I have not ran into your issue. I do not believe it is an TLS issue as it was working a month ago. Double check your user credentials and integration code.

Technically, you do not need to make the 'getZoneInfo' call. All that does is get the proper API endpoint based on your user account.

You can try to bypass 'getZoneInfo' by setting your URL directly. [You can find a list of Autotask endpoints here.](https://ww5.autotask.net/help/Content/AdminSetup/2ExtensionsIntegrations/APIs/WSAPI/General_Topics/Zones_WSDL.htm) I do need to mention, only use this option if you are connecting to a single autotask database, otherwise, you will need to call the getZoneInfo method.

I believe the error is from the following line, as your error is 'Parsing WSDL: Couldn't load' meaning it did make an external call and trying to process. The line you are referring to instantiates the class, but does not make any calls.

You can also uninstall the library and re-install it, if you feel that is necessary.

**Comment by kev22257** [comment]
Thanks for the tips @SirPoot-.  I went through your suggestions and have the same results:

- A new copy of the autotask-php libraries did not affect the behavior at all.
- I commented out the zone code and set $wsdl to the America East WSDL from the page you provided: I get the same error.
- Since there is no authentication required to view the WSDL I'm not convinced the error is credential related: in any case I do not provide them on the zone lookup.

In both cases (before removing the zone logic and after using the specific WSDL for my zone) PHP very clearly fails on the instantiation of the client and not a subsequent call: I believe the parsing of the WSDL XML is done as part of class construction.

Do you have any other ideas?

**Comment by kev22257** [comment]
I made a small PHP test script as follows:

```
$xml = file_get_contents('https://webservices3.autotask.net/atservices/1.6/atws.wsdl');
echo $xml . "\n";
echo strlen($xml). "\n";
```

Sometimes I get the entire WSDL, but the majority of the time I get around 65k of data.  The point at which it stops getting the entire WSDL also varies.  This is definitely confirms this is not a autotask-php issue.  I will update here if I figure it out.

**Comment by kev22257** [comment]
Okay!  We found that the Content-Length header was being corrupted by PHP and causing only part of the WSDL to be downloaded.  This is why the premature end of data error.  We found this was caused by gzip compression used to download the file from Autotask.  Disabling this via the http stream settings on SoapClient did the trick. 

TL;DR ancient PHP problems.

https://stackoverflow.com/questions/61681804/php-file-get-contents-returning-inconsistent-partial-data-for-http-requests
